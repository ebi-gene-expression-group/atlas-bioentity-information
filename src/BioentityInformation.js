import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'

import BioentityProperty from './BioentityProperty'
import Loading from './Loading'

const fetchResponseJson = async (base, endpoint) => {
  const response = await fetch(URI(endpoint, base).toString())
  const responseJson = await response.json()
  return responseJson
}

class BioentityInformation extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      bioentityProperties: [],
      errorMessage: null,
      loading: false
    }
  }

  _fetchAndSetState({atlasUrl, geneId}) {
    const atlasEndpoint = `json/bioentity_information/${geneId}`

    return fetchResponseJson(atlasUrl, atlasEndpoint)
      .then((responseJson) => {
        this.setState({
          bioentityProperties: responseJson,
          errorMessage: null,
          loading: false
        })
      })
      .catch((reason) => {
        this.setState({
          errorMessage: `${reason.name}: ${reason.message}`
        })
      })
  }

  componentDidCatch(error, errorMessage) {
    this.setState({
      errorMessage: errorMessage
    })
  }

  componentDidMount() {
    this.setState({
      loading: true
    })
    return this._fetchAndSetState(this.props)
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.atlasUrl !== this.props.atlasUrl || nextProps.geneId !== this.props.geneId) {
      this.setState({
        bioentityProperties: [],
        loading: true
      })
      this._fetchAndSetState(nextProps)
    }
  }

  render() {
    return (
      <div className={`row`}>
        <div className={`small-12 columns`}>
          <Loading loading={this.state.loading} resourceUrl={this.props.atlasUrl}/>
          <table>
            <tbody>
            {this.state.bioentityProperties.map(function(bioentityProperty){
              return (
                <BioentityProperty
                  key={bioentityProperty.type}
                  {...bioentityProperty} />
              )
            })}
            </tbody>
          </table>
        </div>
      </div>
    )
  }
}

BioentityInformation.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  geneId: PropTypes.string.isRequired
}

export default BioentityInformation
