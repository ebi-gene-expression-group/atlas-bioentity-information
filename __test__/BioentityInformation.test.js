import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import {shallow, mount, render} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import BioentityInformation from '../src/BioentityInformation.js'

Enzyme.configure({ adapter: new Adapter() })

describe(`BioentityInformation`, () => {
  test(`with no data matches snapshot`, () => {
    const tree = renderer.create(<BioentityInformation atlasUrl={``} geneId={``} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test(`should render without throwing an error`, () => {
    expect(mount(<BioentityInformation atlasUrl={``} geneId={``} />)).toHaveLength(1)
  })
})
