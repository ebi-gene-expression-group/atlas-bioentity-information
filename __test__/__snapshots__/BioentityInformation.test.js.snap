// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BioentityInformation with no data matches snapshot 1`] = `
<div
  className="row"
>
  <div
    className="small-12 columns"
  >
    <div
      className="text-center"
    >
      <img
        className="small-centered"
        src="resources/images/loading.gif"
      />
    </div>
    <table>
      <tbody />
    </table>
  </div>
</div>
`;
